assets/model_monitoring/components/src/shared_utilities/span_tree_utils.py

```diff
@@ -69,26 +78,156 @@ def attributes(self) -> str:
+            return self.output_from_attribute()
+        else:
+            events_array: list = json.loads(self.events)
+            if len(events_array) == 0:
+                return self.output_from_attribute()
+            output_event = list(filter(lambda event: event.get("name", None) == APP_TRACES_EVENT_LOG_OUTPUT_KEY,
+                                       events_array))
+            if output_event is not None and len(output_event) > 0:
+                if output_event[0].get("attributes", None) is None:
+                    return None
+                return output_event[0].get("attributes", None).get("payload", None)
+            else:
+                return None
+
+    def output_from_attribute(self) -> str:
+        """Get output from attributes."""
         if self.attributes is None:
             return None  # type: ignore
         attribute_dict: dict = json.loads(self.attributes)
         if attribute_dict is None:
-            return None  # type: ignore
+            return None
         return attribute_dict.get("output", None)
 
+    @property
+    def retrieval_query(self) -> str:
+        """Get promptflow.retrieval.query for retrieval span."""
+        if self.span_type != RETRIEVAL_SPAN_TYPE:
+            return None
+        if self.events is None or len(self.events) == 0:
+            return self.retrieval_query_from_attributes()
+        events_array: list = json.loads(self.events)
+        if len(events_array) == 0:
+            return self.retrieval_query_from_attributes()
+        retrieval_event = list(filter(lambda event: event.get("name", None) ==
+                                      APP_TRACES_EVENT_LOG_RETRIEVAL_QUERY_KEY,
+                                      events_array))
+        if retrieval_event and len(retrieval_event) > 0:
+            if retrieval_event[0].get("attributes", None) is None:
+                return None
+            return retrieval_event[0].get("attributes", None).get("payload", None)
+        else:
+            return None
+
+    @property
+    def retrieval_document(self) -> str:
+        """Get promptflow.retrieval.document for retrieval span."""
+        if self.span_type != RETRIEVAL_SPAN_TYPE:
+            return None
+        if self.events is None or len(self.events) == 0:
+            return self.retrieval_document_from_attributes()
+        events_array: list = json.loads(self.events)
+        if len(events_array) == 0:
+            return self.retrieval_document_from_attributes()
+        retrieval_event = list(filter(lambda event: event.get("name", None) ==
+                                      APP_TRACES_EVENT_LOG_RETRIEVAL_DOCUMENT_KEY,
+                                      events_array))
+        if retrieval_event and len(retrieval_event) > 0:
+            if retrieval_event[0].get("attributes", None) is None:
+                return None
+            return retrieval_event[0].get("attributes", None).get("payload", None)
```

Comment:
json.loads() to return object instead of str

Reply:
Done